File: 135_MaxXORQueries.py
"""
Problem: Maximum XOR with Element from Array (Queries)
Description: Answer queries where each query asks for max XOR of given element
             with elements ≤ given limit.

Approach: Sort queries and array. Build trie incrementally with elements ≤ limit.
"""

from typing import List

class TrieNode:
    def __init__(self):
        self.children = [None, None]

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, num: int):
        node = self.root
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            if node.children[bit] is None:
                node.children[bit] = TrieNode()
            node = node.children[bit]
    
    def findMaxXOR(self, num: int) -> int:
        if self.root.children[0] is None and self.root.children[1] is None:
            return -1
        
        node = self.root
        maxXOR = 0
        
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            oppositeBit = 1 - bit
            
            if node.children[oppositeBit]:
                maxXOR |= (1 << i)
                node = node.children[oppositeBit]
            else:
                node = node.children[bit]
        
        return maxXOR

def maximizeXor(nums: List[int], queries: List[List[int]]) -> List[int]:
    nums.sort()
    
    # Add index to queries and sort by limit
    q = [(queries[i][1], queries[i][0], i) for i in range(len(queries))]
    q.sort()
    
    result = [0] * len(queries)
    trie = Trie()
    idx = 0
    
    for limit, x, originalIndex in q:
        # Add all numbers <= limit to trie
        while idx < len(nums) and nums[idx] <= limit:
            trie.insert(nums[idx])
            idx += 1
        
        result[originalIndex] = trie.findMaxXOR(x)
    
    return result

def main():
    nums = [0, 1, 2, 3, 4]
    queries = [[3, 1], [1, 3], [5, 6]]
    
    result = maximizeXor(nums, queries)
    
    print("Results:", result)  # Output: [3, 3, 7]

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n log n + q log q + (n + q) * 32)
Space Complexity: O(n * 32)
"""

File: 135_MaxXORQueries.cpp
/*
Problem: Maximum XOR with Element from Array (Queries)
Description: Answer queries where each query asks for max XOR of given element
             with elements ≤ given limit.

Approach: Sort queries and array. Build trie incrementally with elements ≤ limit.
*/

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[2];
    
    TrieNode() {
        children[0] = nullptr;
        children[1] = nullptr;
    }
};

class Trie {
    TrieNode* root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(int num) {
        TrieNode* node = root;
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (!node->children[bit])
                node->children[bit] = new TrieNode();
            node = node->children[bit];
        }
    }
    
    int findMaxXOR(int num) {
        if (!root->children[0] && !root->children[1])
            return -1;
        
        TrieNode* node = root;
        int maxXOR = 0;
        
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            int oppositeBit = 1 - bit;
            
            if (node->children[oppositeBit]) {
                maxXOR |= (1 << i);
                node = node->children[oppositeBit];
            } else {
                node = node->children[bit];
            }
        }
        
        return maxXOR;
    }
};

vector<int> maximizeXor(vector<int>& nums, vector<vector<int>>& queries) {
    sort(nums.begin(), nums.end());
    
    // Add index to queries and sort by limit
    vector<tuple<int, int, int>> q;  // {limit, x, originalIndex}
    for (int i = 0; i < queries.size(); i++) {
        q.push_back({queries[i][1], queries[i][0], i});
    }
    sort(q.begin(), q.end());
    
    vector<int> result(queries.size());
    Trie trie;
    int idx = 0;
    
    for (auto [limit, x, originalIndex] : q) {
        // Add all numbers <= limit to trie
        while (idx < nums.size() && nums[idx] <= limit) {
            trie.insert(nums[idx]);
            idx++;
        }
        
        result[originalIndex] = trie.findMaxXOR(x);
    }
    
    return result;
}

int main() {
    vector<int> nums = {0, 1, 2, 3, 4};
    vector<vector<int>> queries = {{3, 1}, {1, 3}, {5, 6}};
    
    vector<int> result = maximizeXor(nums, queries);
    
    cout << "Results: ";
    for (int r : result) {
        cout << r << " ";
    }
    cout << endl;  // Output: 3 3 7
    
    return 0;
}

/*
Time Complexity: O(n log n + q log q + (n + q) * 32)
Space Complexity: O(n * 32)
*/
